runtime/src/account_rent_state.rs

```diff
@@ -0,0 +1,135 @@
+use {
+    enum_iterator::IntoEnumIterator,
+    log::*,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        rent::Rent,
+        transaction::{Result, TransactionError},
+        transaction_context::TransactionContext,
+    },
+};
+
+#[derive(Debug, PartialEq, IntoEnumIterator)]
+pub(crate) enum RentState {
+    Uninitialized, // account.lamports == 0
+    RentPaying,    // 0 < account.lamports < rent-exempt-minimum
+    RentExempt,    // account.lamports >= rent-exempt-minimum
+}
+
+impl RentState {
+    pub(crate) fn from_account(account: &AccountSharedData, rent: &Rent) -> Self {
+        if account.lamports() == 0 {
+            Self::Uninitialized
+        } else if !rent.is_exempt(account.lamports(), account.data().len()) {
+            Self::RentPaying
+        } else {
+            Self::RentExempt
+        }
+    }
+
+    pub(crate) fn transition_allowed_from(&self, pre_rent_state: &RentState) -> bool {
+        // Only a legacy RentPaying account may end in the RentPaying state after message processing
+        !(self == &Self::RentPaying && pre_rent_state != &Self::RentPaying)
+    }
+}
+
+pub(crate) fn submit_rent_state_metrics(pre_rent_state: &RentState, post_rent_state: &RentState) {
+    match (pre_rent_state, post_rent_state) {
+        (&RentState::Uninitialized, &RentState::RentPaying) => {
+            inc_new_counter_info!("rent_paying_err-new_account", 1);
+        }
+        (&RentState::RentPaying, &RentState::RentPaying) => {
+            inc_new_counter_info!("rent_paying_ok-legacy", 1);
+        }
+        (_, &RentState::RentPaying) => {
+            inc_new_counter_info!("rent_paying_err-other", 1);
+        }
+        _ => {}
+    }
+}
+
+pub(crate) fn check_rent_state(
+    pre_rent_state: Option<&RentState>,
+    post_rent_state: Option<&RentState>,
+    transaction_context: &TransactionContext,
+    index: usize,
+) -> Result<()> {
+    if let Some((pre_rent_state, post_rent_state)) = pre_rent_state.zip(post_rent_state) {
+        if !post_rent_state.transition_allowed_from(pre_rent_state) {
+            debug!(
+                "Account {:?} not rent exempt, state {:?}",
+                transaction_context.get_key_of_account_at_index(index),
+                transaction_context.get_account_at_index(index).borrow(),
+            );
+            submit_rent_state_metrics(pre_rent_state, post_rent_state);
```

Comment:
This can record metrics for valid legacy -> legacy rent paying transitions but that path won't ever be hit right? Should this be called regardless of whether the transition was valid?

Reply:
Oh, too true! Thanks for catching that